File: tests/DesktopGrass.Win2D.Tests/SimTests/FireflyTests.cs
// FireflyTests.cs - §17.7 ambient Firefly tests.

using System;
using System.Collections.Generic;
using System.Linq;
using DesktopGrass.Win2D;
using Xunit;

namespace DesktopGrass.Win2D.Tests.SimTests;

public class FireflyTests
{
    private const double Monitor1920 = 1920.0;

    private static Sim BuildSim(ulong seed = Constants.CANONICAL_TEST_SEED)
    {
        var sim = new Sim
        {
            Blades = Sim.GenerateBlades(seed, Monitor1920, Constants.DEFAULT_DENSITY),
            WindowHeight = Constants.STRIP_HEIGHT + Constants.HEADROOM,
            GroundY = Constants.STRIP_HEIGHT + Constants.HEADROOM,
        };
        sim.ResetAmbientGusts(seed, Monitor1920);
        sim.ResetEntities(seed);
        return sim;
    }

    private static Sim BuildGrassSim(ulong seed = Constants.CANONICAL_TEST_SEED)
    {
        var sim = BuildSim(seed);
        sim.SetScene(Scene.Grass);
        return sim;
    }

    private static int CountKind(Sim sim, EntityKind kind) => sim.Entities.Count(e => e.Kind == kind);

    private static int PrngCount(ref Prng side, int minCount, int maxCount)
    {
        double draw = side.Uniform(minCount, maxCount + 1);
        int count = (int)Math.Floor(draw);
        if (count < minCount) count = minCount;
        if (count > maxCount) count = maxCount;
        return count;
    }

    [Fact]
    public void FireflyConstantsArePinnedToSpecValues()
    {
        Assert.Equal(3, Constants.FIREFLY_COUNT_MIN);
        Assert.Equal(6, Constants.FIREFLY_COUNT_MAX);
        Assert.Equal(4.0, Constants.FIREFLY_DRIFT_SPEED_MIN);
        Assert.Equal(10.0, Constants.FIREFLY_DRIFT_SPEED_MAX);
        Assert.Equal(1.2, Constants.FIREFLY_BODY_RADIUS);
        Assert.Equal(5.0, Constants.FIREFLY_GLOW_RADIUS);
        Assert.Equal(1.4, Constants.FIREFLY_BLINK_PERIOD_MIN);
        Assert.Equal(2.6, Constants.FIREFLY_BLINK_PERIOD_MAX);
        Assert.Equal(0.55, Constants.FIREFLY_BLINK_DUTY);
        Assert.Equal(0.30, Constants.FIREFLY_BLINK_FADE);
        Assert.Equal(0.4, Constants.FIREFLY_DRIFT_FREQ_X);
        Assert.Equal(0.6, Constants.FIREFLY_DRIFT_FREQ_Y);
        Assert.Equal(0.6, Constants.FIREFLY_DRIFT_AMP_X);
        Assert.Equal(8.0, Constants.FIREFLY_DRIFT_AMP_Y);
        Assert.Equal(8.0, Constants.FIREFLY_ALTITUDE_MIN);
        Assert.Equal(55.0, Constants.FIREFLY_ALTITUDE_MAX);
        Assert.Equal(0xFFFFEE88u, Constants.FIREFLY_BODY_COLOR);
        Assert.Equal(0xEEDD66u, Constants.FIREFLY_GLOW_COLOR_RGB);
        Assert.Equal(110, Constants.FIREFLY_GLOW_ALPHA_MAX);
        Assert.Equal(255, Constants.FIREFLY_BODY_ALPHA_MAX);
        Assert.Equal(20, Constants.FIREFLY_NIGHT_START_HOUR);
        Assert.Equal(6, Constants.FIREFLY_NIGHT_END_HOUR);
        Assert.Equal(1, Constants.FIREFLY_FADE_DURATION_HOUR);
        Assert.Equal(0xF13EF1E7777ul, Constants.FIREFLY_PRNG_SALT);
    }

    [Fact]
    public void GrassGenerationProducesFireflyCountInRange()
    {
        for (ulong i = 0; i < 128; i++)
        {
            ulong seed = unchecked(Constants.CANONICAL_TEST_SEED + i * 0x9E3779B97F4A7C15UL);
            var sim = BuildGrassSim(seed);
            Assert.InRange(CountKind(sim, EntityKind.Firefly), Constants.FIREFLY_COUNT_MIN, Constants.FIREFLY_COUNT_MAX);
        }
    }

    [Fact]
    public void FirefliesAreGrassSceneOnly()
    {
        var sim = BuildSim();
        sim.SetScene(Scene.Desert);
        Assert.Equal(0, CountKind(sim, EntityKind.Firefly));
        sim.SetScene(Scene.Winter);
        Assert.Equal(0, CountKind(sim, EntityKind.Firefly));
    }

    [Fact]
    public void GeneratedFirefliesHaveSpeedAltitudeAndBlinkPeriodRanges()
    {
        var sim = BuildGrassSim();
        foreach (var e in sim.Entities.Where(e => e.Kind == EntityKind.Firefly))
        {
            Assert.InRange(e.BaseSpeed, Constants.FIREFLY_DRIFT_SPEED_MIN, Constants.FIREFLY_DRIFT_SPEED_MAX);
            Assert.InRange(e.AltitudeAnchor, Constants.FIREFLY_ALTITUDE_MIN, Constants.FIREFLY_ALTITUDE_MAX);
            Assert.InRange(e.BlinkPeriod, Constants.FIREFLY_BLINK_PERIOD_MIN, Constants.FIREFLY_BLINK_PERIOD_MAX);
        }
    }

    [Fact]
    public void FireflyPrngDrawOrderMatchesSideStream()
    {
        var side = Prng.Init(Constants.CANONICAL_TEST_SEED ^ Constants.FIREFLY_PRNG_SALT);
        var sim = BuildGrassSim();

        int expectedCount = PrngCount(ref side, Constants.FIREFLY_COUNT_MIN, Constants.FIREFLY_COUNT_MAX);
        Assert.Equal(expectedCount, CountKind(sim, EntityKind.Firefly));

        int seen = 0;
        foreach (var e in sim.Entities.Where(e => e.Kind == EntityKind.Firefly))
        {
            double xFrac = side.Uniform(0.0, 1.0);
            double yFrac = side.Uniform(0.0, 1.0);
            ulong vxSign = side.NextU64() & 1UL;
            double expectedDir = vxSign != 0UL ? 1.0 : -1.0;
            double expectedSpeed = side.Uniform(Constants.FIREFLY_DRIFT_SPEED_MIN, Constants.FIREFLY_DRIFT_SPEED_MAX);
            double expectedBlinkPeriod = side.Uniform(Constants.FIREFLY_BLINK_PERIOD_MIN, Constants.FIREFLY_BLINK_PERIOD_MAX);
            double expectedBlinkPhase = side.Uniform(0.0, 1.0);
            double expectedPhaseY = side.Uniform(0.0, 2.0 * Math.PI);
            double expectedPhaseX = side.Uniform(0.0, 2.0 * Math.PI);
            double expectedAltitude = Constants.FIREFLY_ALTITUDE_MIN
                + yFrac * (Constants.FIREFLY_ALTITUDE_MAX - Constants.FIREFLY_ALTITUDE_MIN);
            double expectedVx = expectedDir * expectedSpeed
                * (1.0 + Constants.FIREFLY_DRIFT_AMP_X * Math.Sin(expectedPhaseX));

            Assert.Equal(xFrac * Monitor1920, e.X, 9);
            Assert.Equal(expectedAltitude, e.AltitudeAnchor, 9);
            Assert.Equal(expectedSpeed, e.BaseSpeed, 9);
            Assert.Equal(expectedBlinkPeriod, e.BlinkPeriod, 9);
            Assert.Equal(expectedBlinkPhase, e.BlinkPhase, 9);
            Assert.Equal(expectedVx, e.Vx, 9);
            Assert.Equal(expectedPhaseY, e.PhaseY, 9);
            Assert.Equal(expectedPhaseX, e.PhaseX, 9);
            seen++;
        }
        Assert.Equal(expectedCount, seen);
    }

    [Fact]
    public void FireflyEdgeWrapPreservesAltitudeAnchor()
    {
        var sim = BuildGrassSim();
        int idx = sim.Entities.FindIndex(e => e.Kind == EntityKind.Firefly);
        Assert.True(idx >= 0);
        double margin = Constants.FIREFLY_GLOW_RADIUS;
        var e = sim.Entities[idx];
        e.X = Monitor1920 + margin + 1.0;
        e.Vx = Math.Abs(e.Vx);
        double altitude = e.AltitudeAnchor;
        sim.Entities[idx] = e;
        sim.CurrentScene = Scene.Desert;

        sim.TickEntities(0.016);

        Assert.Equal(-margin, sim.Entities[idx].X, 9);
        Assert.Equal(altitude, sim.Entities[idx].AltitudeAnchor, 9);
    }

    [Fact]
    public void FirefliesDoNotInteractWithCutsOrPets()
    {
        var sim = BuildSim();
        sim.Entities.Clear();
        var firefly = new Entity
        {
            Kind = EntityKind.Firefly,
            X = 500.0,
            Y = sim.WindowHeight - Constants.STRIP_HEIGHT - 5.0,
            Vx = Constants.FIREFLY_DRIFT_SPEED_MIN,
            BaseSpeed = Constants.FIREFLY_DRIFT_SPEED_MIN,
            AltitudeAnchor = Constants.FIREFLY_ALTITUDE_MIN,
            BlinkPeriod = Constants.FIREFLY_BLINK_PERIOD_MIN,
            Lifetime = -1.0,
        };
        sim.Entities.Add(firefly);
        sim.Entities.Add(new Entity
        {
            Kind = EntityKind.Sheep,
            X = firefly.X,
            Y = sim.WindowHeight - Constants.SHEEP_BODY_HEIGHT - Constants.SHEEP_LEG_LENGTH,
            Vx = Constants.SHEEP_WALK_SPEED_MIN,
            State = Constants.SHEEP_STATE_WALKING,
            StateTimer = 10.0,
        });

        sim.ApplyClick(firefly.X, firefly.Y, 0.0);

        Assert.Equal(EntityKind.Firefly, sim.Entities[0].Kind);
        Assert.Equal(Constants.FIREFLY_DRIFT_SPEED_MIN, sim.Entities[0].BaseSpeed, 9);
        Assert.Equal(Constants.SHEEP_STATE_WALKING, sim.Entities[1].State);
        Assert.All(sim.Blades, b => Assert.True(b.CutAnimStart < 0.0));
    }

    [Fact]
    public void FireflyBlinkBrightnessHasOnAndOffPhases()
    {
        const double period = 2.0;
        Assert.Equal(1.0, Constants.FireflyBlinkBrightness(period * 0.25, period, 0.0), 9);
        Assert.Equal(0.0, Constants.FireflyBlinkBrightness(period * 0.80, period, 0.0), 9);
    }

    [Fact]
    public void FireflyPhasesDecorrelateVisibleBrightness()
    {
        const double period = 2.0;
        double[] phases = { 0.0, 0.0375, 0.075, 0.1125, 0.25, 0.80 };
        var distinct = new List<double>();
        foreach (double phase in phases)
        {
            double brightness = Constants.FireflyBlinkBrightness(0.0, period, phase);
            if (!distinct.Any(existing => Math.Abs(existing - brightness) < 1e-6))
            {
                distinct.Add(brightness);
            }
        }
        Assert.True(distinct.Count >= 4);
    }

    [Fact]
    public void FireflyFadeIsNightOnlyWithDuskAndDawnRamps()
    {
        Assert.Equal(1.0, Constants.FireflyFade(0.0), 9);
        Assert.Equal(0.0, Constants.FireflyFade(12.0), 9);
        Assert.Equal(0.5, Constants.FireflyFade(19.5), 9);
        Assert.Equal(0.5, Constants.FireflyFade(6.5), 9);
    }
}

File: tests/DesktopGrass.Native.Tests/src/firefly_tests.cpp
// firefly_tests.cpp - §17.7 ambient Firefly tests.

#include "TestHelpers.h"
#include "Sim.h"

#include <algorithm>
#include <cmath>
#include <vector>

using namespace desktopgrass;

namespace {
constexpr double Monitor1920 = 1920.0;

int count_kind(const Sim& sim, EntityKind kind) {
    return static_cast<int>(std::count_if(sim.entities.begin(), sim.entities.end(),
        [kind](const Entity& e) { return e.kind == kind; }));
}

Sim build_grass_sim(uint64_t seed = CANONICAL_TEST_SEED) {
    Sim sim = sim_init(seed, Monitor1920, DEFAULT_DENSITY);
    sim_set_scene(sim, Scene::Grass);
    return sim;
}

int prng_count(Prng& side, int minCount, int maxCount) {
    const double draw = prng_uniform(side, static_cast<double>(minCount), static_cast<double>(maxCount + 1));
    int count = static_cast<int>(std::floor(draw));
    if (count < minCount) count = minCount;
    if (count > maxCount) count = maxCount;
    return count;
}
} // namespace

using namespace Microsoft::VisualStudio::CppUnitTestFramework;

namespace DesktopGrassNativeTests
{
TEST_CLASS(FireflyTests)
{
public:
TEST_METHOD(FireflyConstantsArePinnedToSpecValues) {
    Assert::IsTrue(FIREFLY_COUNT_MIN == 3);
    Assert::IsTrue(FIREFLY_COUNT_MAX == 6);
    Assert::IsTrue(FIREFLY_DRIFT_SPEED_MIN == Near(4.0));
    Assert::IsTrue(FIREFLY_DRIFT_SPEED_MAX == Near(10.0));
    Assert::IsTrue(FIREFLY_BODY_RADIUS == Near(1.2));
    Assert::IsTrue(FIREFLY_GLOW_RADIUS == Near(5.0));
    Assert::IsTrue(FIREFLY_BLINK_PERIOD_MIN == Near(1.4));
    Assert::IsTrue(FIREFLY_BLINK_PERIOD_MAX == Near(2.6));
    Assert::IsTrue(FIREFLY_BLINK_DUTY == Near(0.55));
    Assert::IsTrue(FIREFLY_BLINK_FADE == Near(0.30));
    Assert::IsTrue(FIREFLY_DRIFT_FREQ_X == Near(0.4));
    Assert::IsTrue(FIREFLY_DRIFT_FREQ_Y == Near(0.6));
    Assert::IsTrue(FIREFLY_DRIFT_AMP_X == Near(0.6));
    Assert::IsTrue(FIREFLY_DRIFT_AMP_Y == Near(8.0));
    Assert::IsTrue(FIREFLY_ALTITUDE_MIN == Near(8.0));
    Assert::IsTrue(FIREFLY_ALTITUDE_MAX == Near(55.0));
    Assert::IsTrue(FIREFLY_BODY_COLOR == 0xFFFFEE88u);
    Assert::IsTrue(FIREFLY_GLOW_COLOR_RGB == 0xEEDD66u);
    Assert::IsTrue(FIREFLY_GLOW_ALPHA_MAX == 110);
    Assert::IsTrue(FIREFLY_BODY_ALPHA_MAX == 255);
    Assert::IsTrue(FIREFLY_PRNG_SALT == 0xF13EF1E7777ull);
}

TEST_METHOD(GrassGenerationProducesFireflyCountInRange) {
    for (uint64_t i = 0; i < 128; ++i) {
        const uint64_t seed = CANONICAL_TEST_SEED + i * 0x9E3779B97F4A7C15ull;
        Sim sim = build_grass_sim(seed);
        Assert::IsTrue(count_kind(sim, EntityKind::Firefly) >= FIREFLY_COUNT_MIN);
        Assert::IsTrue(count_kind(sim, EntityKind::Firefly) <= FIREFLY_COUNT_MAX);
    }
}

TEST_METHOD(FirefliesAreGrassSceneOnly) {
    Sim sim = sim_init(CANONICAL_TEST_SEED, Monitor1920, DEFAULT_DENSITY);
    sim_set_scene(sim, Scene::Desert);
    Assert::IsTrue(count_kind(sim, EntityKind::Firefly) == 0);
    sim_set_scene(sim, Scene::Winter);
    Assert::IsTrue(count_kind(sim, EntityKind::Firefly) == 0);
}

TEST_METHOD(GeneratedFirefliesHaveSpeedAltitudeAndBlinkPeriodRanges) {
    Sim sim = build_grass_sim();
    for (const Entity& e : sim.entities) {
        if (e.kind != EntityKind::Firefly) continue;
        Assert::IsTrue(e.baseSpeed >= FIREFLY_DRIFT_SPEED_MIN);
        Assert::IsTrue(e.baseSpeed <  FIREFLY_DRIFT_SPEED_MAX);
        Assert::IsTrue(e.altitudeAnchor >= FIREFLY_ALTITUDE_MIN);
        Assert::IsTrue(e.altitudeAnchor <  FIREFLY_ALTITUDE_MAX);
        Assert::IsTrue(e.blinkPeriod >= FIREFLY_BLINK_PERIOD_MIN);
        Assert::IsTrue(e.blinkPeriod <  FIREFLY_BLINK_PERIOD_MAX);
    }
}

TEST_METHOD(FireflyPRNGDrawOrderMatchesSideStream) {
    Prng side;
    prng_init(side, CANONICAL_TEST_SEED ^ FIREFLY_PRNG_SALT);
    Sim sim = build_grass_sim();

    const int expectedCount = prng_count(side, FIREFLY_COUNT_MIN, FIREFLY_COUNT_MAX);
    Assert::IsTrue(count_kind(sim, EntityKind::Firefly) == expectedCount);

    int seen = 0;
    for (const Entity& e : sim.entities) {
        if (e.kind != EntityKind::Firefly) continue;
        const double xFrac = prng_uniform(side, 0.0, 1.0);
        const double yFrac = prng_uniform(side, 0.0, 1.0);
        const uint64_t vxSign = prng_next_u64(side) & 1ull;
        const double expectedDir = vxSign != 0ull ? 1.0 : -1.0;
        const double expectedSpeed = prng_uniform(side, FIREFLY_DRIFT_SPEED_MIN, FIREFLY_DRIFT_SPEED_MAX);
        const double expectedBlinkPeriod = prng_uniform(side, FIREFLY_BLINK_PERIOD_MIN, FIREFLY_BLINK_PERIOD_MAX);
        const double expectedBlinkPhase = prng_uniform(side, 0.0, 1.0);
        const double expectedPhaseY = prng_uniform(side, 0.0, 2.0 * 3.14159265358979323846);
        const double expectedPhaseX = prng_uniform(side, 0.0, 2.0 * 3.14159265358979323846);
        const double expectedAltitude = FIREFLY_ALTITUDE_MIN + yFrac * (FIREFLY_ALTITUDE_MAX - FIREFLY_ALTITUDE_MIN);
        const double expectedVx = expectedDir * expectedSpeed * (1.0 + FIREFLY_DRIFT_AMP_X * std::sin(expectedPhaseX));

        Assert::IsTrue(e.x == Near(xFrac * Monitor1920));
        Assert::IsTrue(e.altitudeAnchor == Near(expectedAltitude));
        Assert::IsTrue(e.baseSpeed == Near(expectedSpeed));
        Assert::IsTrue(e.blinkPeriod == Near(expectedBlinkPeriod));
        Assert::IsTrue(e.blinkPhase == Near(expectedBlinkPhase));
        Assert::IsTrue(e.vx == Near(expectedVx));
        Assert::IsTrue(e.phaseY == Near(expectedPhaseY));
        Assert::IsTrue(e.phaseX == Near(expectedPhaseX));
        ++seen;
    }
    Assert::IsTrue(seen == expectedCount);
}

TEST_METHOD(FireflyEdgeWrapPreservesAltitudeAnchor) {
    Sim sim = build_grass_sim();
    auto it = std::find_if(sim.entities.begin(), sim.entities.end(), [](const Entity& e) { return e.kind == EntityKind::Firefly; });
    Assert::IsTrue(it != sim.entities.end());
    const double margin = FIREFLY_GLOW_RADIUS;
    it->x = Monitor1920 + margin + 1.0;
    it->vx = std::abs(it->vx);
    const double altitude = it->altitudeAnchor;
    sim.currentScene = Scene::Desert;

    sim_tick_entities(sim, 0.016);

    Assert::IsTrue(it->x == Near(-margin));
    Assert::IsTrue(it->altitudeAnchor == Near(altitude));
}

TEST_METHOD(FirefliesDoNotInteractWithCutsOrPets) {
    Sim sim = sim_init(CANONICAL_TEST_SEED, Monitor1920, DEFAULT_DENSITY);
    sim.entities.clear();
    Entity firefly{};
    firefly.kind = EntityKind::Firefly;
    firefly.x = 500.0;
    firefly.y = sim.windowHeight - STRIP_HEIGHT - 5.0;
    firefly.vx = FIREFLY_DRIFT_SPEED_MIN;
    firefly.baseSpeed = FIREFLY_DRIFT_SPEED_MIN;
    firefly.altitudeAnchor = FIREFLY_ALTITUDE_MIN;
    firefly.blinkPeriod = FIREFLY_BLINK_PERIOD_MIN;
    firefly.lifetime = -1.0;
    sim.entities.push_back(firefly);
    Entity sheep{};
    sheep.kind = EntityKind::Sheep;
    sheep.x = firefly.x;
    sheep.y = sim.windowHeight - SHEEP_BODY_HEIGHT - SHEEP_LEG_LENGTH;
    sheep.vx = SHEEP_WALK_SPEED_MIN;
    sheep.state = SHEEP_STATE_WALKING;
    sheep.stateTimer = 10.0;
    sim.entities.push_back(sheep);

    InputEvent ev{};
    ev.type = EventType::Click;
    ev.x = firefly.x;
    ev.y = firefly.y;
    sim_apply_click(sim, ev);

    Assert::IsTrue(sim.entities[0].kind == EntityKind::Firefly);
    Assert::IsTrue(sim.entities[0].baseSpeed == Near(FIREFLY_DRIFT_SPEED_MIN));
    Assert::IsTrue(sim.entities[1].state == SHEEP_STATE_WALKING);
    for (const Blade& b : sim.blades) Assert::IsTrue(b.cutAnimStart < 0.0);
}

TEST_METHOD(FireflyBlinkBrightnessHasOnAndOffPhases) {
    const double period = 2.0;
    Assert::IsTrue(firefly_blink_brightness(period * 0.25, period, 0.0) == Near(1.0));
    Assert::IsTrue(firefly_blink_brightness(period * 0.80, period, 0.0) == Near(0.0));
}

TEST_METHOD(FireflyPhasesDecorrelateVisibleBrightness) {
    const double period = 2.0;
    const double phases[] = { 0.0, 0.0375, 0.075, 0.1125, 0.25, 0.80 };
    std::vector<double> distinct;
    for (double phase : phases) {
        const double b = firefly_blink_brightness(0.0, period, phase);
        bool seen = false;
        for (double existing : distinct) {
            if (std::fabs(existing - b) < 1e-6) { seen = true; break; }
        }
        if (!seen) distinct.push_back(b);
    }
    Assert::IsTrue(distinct.size() >= 4);
}
};
}
